datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -522,6 +522,13 @@ fn date_bin_impl(
 
     let (stride, stride_fn) = stride.bin_fn();
 
+    // Return error if stride is 0
+    if stride == 0 {
```

Comment:
The error says the stride must be positive and non zero, but this check is only for `0`

It seems like the main branch supports negative intervals:
```
❯ SELECT DATE_BIN(INTERVAL '-15 minutes', TIMESTAMP '2022-08-03 14:38:50Z', TIMESTAMP '1970-01-01T00:00:00Z');
+-----------------------------------------------------------------------------------------------------------------+
| datebin(IntervalMonthDayNano("18446743173709551616"),Utf8("2022-08-03 14:38:50Z"),Utf8("1970-01-01T00:00:00Z")) |
+-----------------------------------------------------------------------------------------------------------------+
| 2022-08-03T14:30:00                                                                                             |
+-----------------------------------------------------------------------------------------------------------------+
1 row in set. Query took 0.001 seconds.
```

Thus, given what @comphead  found in https://github.com/apache/arrow-datafusion/pull/6522#pullrequestreview-1456448347

I think we should either:
1. Update the error to say that the DATE_BIN stride must be non zero
2. Update the code to match the error (prevent negative strides) and then add test coverage for negative intervals

Reply:
Since I do not know if the negative stride is therefor a specific purpose and someone may use it, I chose not to change its behavior. I go with the option 1 to only fix the panic and have change the error message to "DATE_BIN stride must be non-zero"

Reply:
> Thanks @comphead and @alamb, I have addressed your comments.
> 
> @comphead : we supported interval `month` and `year` interval in [this PR](https://github.com/apache/arrow-datafusion/pull/5982)

Thanks I'll check that, PG doesn't support strides as  months and larger, need to check what is the reason for that, as we consider PG behavior  as an example

Reply:
Using months as strides is significantly more complicated because they are not a fixed size (sometimes 28 days, sometimes 29, 30 or 31, etc)

Binning by months / years was important to us in IOx so @NGA-TRAN  implemented that logic for https://github.com/apache/arrow-datafusion/issues/5689